### PR TITLE
Add schema v2 adapt workflow

### PR DESCRIPTION
**Pull request by anmorgunov**
why: schema v2 adapters now share a common `iter_raw_routes` / `cast` seam, but there was no workflow API implementing the planned `adapt_route`, `adapt_routes`, and `adapt_candidates` boundary. This adds the clean workflow layer needed before collect/ingest work.

what changed: added `retrocast.v2.workflow.adapt` with route-first adaptation, candidate adaptation that preserves failed rank slots, target-hint handling from `RawRouteEntry`, and failure conversion into v2 `FailureRecord`. Added focused workflow tests using a small in-test adapter.

risk: the important boundary is extraction vs candidate casting. Extraction/schema failures from `iter_raw_routes` propagate because there is no candidate slot yet. Per-entry `cast` failures become failed `Candidate`s for benchmarking accounting. `adapt_routes` remains route-first and filters failed casts.

verification: `uv run pytest tests/v2`; `uv run ruff check src/retrocast/v2/workflow tests/v2/workflow`; `uv run ruff format --check src/retrocast/v2/workflow tests/v2/workflow` after formatting.

follow-ups: collect and ingest workflow APIs are still intentionally left for the next schema-v2 slice.

---
<a href="https://app.blacksmith.sh/ischemist/codesmith/project-procrustes/pr/109"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1782582101&installation_id=125602297&pr_number=109&repository=ischemist%2Fproject-procrustes&return_to=https%3A%2F%2Fgithub.com%2Fischemist%2Fproject-procrustes%2Fpull%2F109&signature=42210b84bcbf0b665aa179b6a06c5b57d5f71cdb518f6a09305314a382f506c7"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

* **New Features**
  * Added workflow adaptation functionality to convert planner outputs into domain models with robust error handling and multiple adaptation modes.

* **Tests**
  * Added comprehensive test suite for the workflow adaptation layer, covering successful conversions, failure handling, and error propagation scenarios.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ischemist/project-procrustes/pull/109?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

<h3>Greptile Summary</h3>

Introduces the `retrocast.v2.workflow.adapt` module with three public functions (`adapt_route`, `adapt_routes`, `adapt_candidates`) that bridge the v2 adapter protocol to canonical `Route`/`Candidate` domain objects, plus a focused test suite backed by a small in-test adapter.

- `adapt_route` / `adapt_routes` provide simple single-route and batch adaptation with silent failure filtering; `adapt_candidates` preserves per-rank failure slots as `FailureRecord` objects, making cast errors visible for benchmarking accounting.
- `_target_from_entry` resolves per-entry target hints, and `_failure_from_exception` normalises `AdapterError`, `ChemError`, and pydantic `ValidationError` into a stable `FailureRecord` shape; the extraction-vs-cast error boundary is intentional and tested.

<h3>Confidence Score: 4/5</h3>

The new workflow module is a thin, well-structured layer over the existing adapter protocol; the logic is straightforward and the test coverage is solid.

Two edge cases worth attention: `Target(...)` construction in `_target_from_entry` sits outside both the local ChemError guard and the per-entry try/except in `adapt_candidates`, so a pydantic ValidationError there would escape as an unhandled exception rather than a FailureRecord. And the `fallback_rank` counter increments on every iteration including entries that already supply `source_order`, which can produce duplicate ranks when the two are mixed. Neither is triggered by the current tests or typical adapter usage, but both are latent misbehaviours that could surface with unusual adapter implementations.

src/retrocast/v2/workflow/adapt.py — specifically `_target_from_entry` and the rank-assignment logic in `adapt_candidates`

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/retrocast/v2/workflow/adapt.py | Core adaptation layer: adapt_route, adapt_routes, adapt_candidates plus two private helpers. Latent ValidationError escape in _target_from_entry and a rank-collision edge case in adapt_candidates under mixed source_order presence. |
| src/retrocast/v2/workflow/__init__.py | Package init re-exporting adapt_candidates, adapt_route, adapt_routes — straightforward and correct. |
| tests/v2/workflow/test_adapt.py | Focused test suite with a minimal in-test adapter covering success paths, filtering, rank preservation, failure-record identity, prune mode, and extraction-failure propagation; good coverage of the intended boundaries. |
| tests/v2/workflow/__init__.py | Empty test package marker — no changes. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Caller
    participant adapt_candidates
    participant Adapter
    participant _target_from_entry
    participant _failure_from_exception

    Caller->>adapt_candidates: raw_payload, adapter, mode, target
    adapt_candidates->>Adapter: iter_raw_routes(raw_payload)
    loop for each RawRouteEntry
        Adapter-->>adapt_candidates: entry (or raises propagates)
        adapt_candidates->>_target_from_entry: entry (if target is None)
        _target_from_entry-->>adapt_candidates: "Target | None"
        adapt_candidates->>Adapter: cast(entry.payload, mode, entry_target)
        alt cast succeeds
            Adapter-->>adapt_candidates: Route
            adapt_candidates->>adapt_candidates: "append Candidate(rank, route=Route)"
        else "AdapterError | ChemError | ValidationError"
            Adapter-->>adapt_candidates: raises exc
            adapt_candidates->>_failure_from_exception: "exc, target=entry_target, entry=entry"
            _failure_from_exception-->>adapt_candidates: FailureRecord
            adapt_candidates->>adapt_candidates: "append Candidate(rank, failure=FailureRecord)"
        end
    end
    adapt_candidates-->>Caller: list[Candidate]
```

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 2 code review issues. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 2
src/retrocast/v2/workflow/adapt.py:66-77
**`Target` construction escapes the ChemError guard**

`Target(...)` is called after the `try/except ChemError` block exits. `Target._validate_identity` internally calls `get_inchi_key(self.smiles)` and wraps any `ChemError` as a `ValueError`, which Pydantic re-raises as a `ValidationError`. If that path were ever hit, the `ValidationError` would escape `_target_from_entry` entirely and — since `_target_from_entry` is invoked before the per-entry `try` block in `adapt_candidates` — would propagate out of the loop as an unhandled error rather than being converted to a `FailureRecord`. Moving the `Target(...)` call inside the `try` block (and broadening the `except` to include `ValidationError`) would close this gap.

### Issue 2 of 2
src/retrocast/v2/workflow/adapt.py:57-58
**Rank collision when `source_order` is partially set**

`fallback_rank` increments on every iteration, including iterations where `entry.source_order` is used instead. So if entry 1 has `source_order=None` (→ rank 1) and entry 2 has `source_order=1` (→ rank 1), both candidates receive rank 1. The counter needs to be independent of whether `source_order` is present — either by tracking the next fallback position separately, or by asserting that all entries in a batch are consistently ranked.

```suggestion
    fallback_rank = 0
    for entry in adapter.iter_raw_routes(raw_payload, source_key=source_key):
        fallback_rank += 1
        rank = entry.source_order if entry.source_order is not None else fallback_rank
```

`````

</details>

<sub>Reviews (1): Last reviewed commit: ["Simplify v2 candidate rank fallback"](https://github.com/ischemist/project-procrustes/commit/cf99e6c547b38cbc1e1876929ba7c8217ad4bc6e) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=34415076)</sub>

> Greptile also left **2 inline comments** on this PR.

### CODE REVIEW

File: src/retrocast/v2/workflow/adapt.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    try:
+        smiles = canonicalize_smiles(entry.target_hint_smiles)
+        inchikey = get_inchi_key(smiles)
+    except ChemError:
+        return None
+    return Target(
+        id=entry.target_hint_id or entry.source_key or smiles,
+        smiles=SmilesStr(smiles),
+        inchikey=InChIKeyStr(inchikey),
+    )
```
[the remark this answers is not in the document]
Comment (anmorgunov) [reply src/retrocast/v2/workflow/adapt.py:84]: good point, but i'm eye-ing an eventual move of inchikey from something users have to provide during construction to derived property.
